File: src/zros_cpp/service_discovery.cc
//
// Created by pairs on 7/13/18.
//

#include <zros/service_discovery.h>
#include <zros/error.h>
#include "impl/service_discovery_impl.h"

namespace zros {
    ServiceDiscovery::ServiceDiscovery(const std::string &master_address, const std::string &agent_address) {
        impl_ = std::make_shared<ServiceDiscoveryImpl>(master_address, agent_address);
    }

    void ServiceDiscovery::init() {
        if (!impl_->isConnectedToMaster()) {
            throw initialize_error("Cannot connect to Master");
        }
    }

    void ServiceDiscovery::spin() {
        impl_->spin();
    }

    ServiceDiscovery::ServiceDiscovery() {
        if (!impl_) {
            throw initialize_error("please init service discovery first");
        }
    }

    bool ServiceDiscovery::addServiceServer(const std::shared_ptr<IServiceServer> server) {

        return impl_->addServiceServer(server);
    }

    bool ServiceDiscovery::addServiceClient(const std::shared_ptr<IServiceClient> client) {
        return impl_->addServiceClient(client);
    }

    bool ServiceDiscovery::addPublisher(const std::shared_ptr<IPublisher> publisher) {
        return impl_->addPublisher(publisher);
    }

    bool ServiceDiscovery::addSubscriber(const std::shared_ptr<ISubscriber> subscriber) {
        return impl_->addSubscriber(subscriber);
    }

    void ServiceDiscovery::set_register_service_server_cb(DealRegisterServiceServerCb cb) {
        impl_->set_register_service_server_cb(cb);
    }

    void ServiceDiscovery::set_unregister_service_server_cb(DealRegisterServiceServerCb cb) {
        impl_->set_unregister_service_server_cb(cb);
    }

    void ServiceDiscovery::set_unregister_publisher_cb(DealRegisterPublisherCb cb) {

    }

    void ServiceDiscovery::set_register_publisher_cb(DealRegisterPublisherCb cb) {
        impl_->set_register_publisher_cb(cb);
    }

}

File: src/zros_python/service_discovery.py
# -*- coding: utf-8 -*-

import logging
import zros_python.impl.service_discovery_impl as impl


class ServiceDiscovery(object):

    def __init__(self, master_address, agent_address):
        self._impl = impl.ServiceDiscoveryImpl(master_address, agent_address)
        if self._impl.is_connected_master() is False:
            raise Exception(u'connected master failed')

    def spin(self):
        self._impl.spin()

    def add_service_server(self, server):
        return self._impl.add_service_server(server)

    def add_service_client(self, client):
        return self._impl.add_service_client(client)

    def add_publisher(self, publisher):
        return self._impl.add_publisher(publisher)

    def add_subscriber(self, subscriber):
        return self._impl.add_subscriber(subscriber)

    def set_register_service_server_cb(self, callback):
        return self._impl.set_register_service_server_cb(callback)

    def set_unregister_service_server_cb(self, callback):
        return self.set_unregister_service_server_cb(callback)

    def set_register_publisher_cb(self, callback):
        return self.set_register_publisher_cb(callback)

    def set_unregister_publisher_cb(self, callback):
        return self.set_unregister_publisher_cb(callback)


if __name__ == '__main__':
    logging.basicConfig(format=u'[%(name)s - %(levelname)s]: %(message)s',
                        level=logging.NOTSET)
    sd = ServiceDiscovery(u'localhost:23333', u'[::]:')
    sd.spin()
